### PR TITLE
Fix test_settlement

### LINKED ISSUE #390: fix `raiden/tests/integration/test_settlement.py`
## Problem Definition

`test_settlement.py::test_settlement` currently fails. In reality it doesn't even properly close the channel.

## Solution

Debug the `close` transaction and see where it throws.

## Tasklist

- [ ] connect the browsersolidity debugger
- [ ] fix the setup of the channel or the `close` call
- [ ] remove `@pytest.mark.skipif` to avoid regressions!

### PR DESCRIPTION
**Pull request by konradkonrad**
This fixes the automatic settlement as tested in `tests/integration/test_settlement.py::test_settlement` and allows to remove the pytest skip on travis-CI. Due to the applied fixes a couple of other tests that apparently adjusted to the formerly broken behavior needed changes, too. 

There were a couple of issues that lead us to start skipping the test and I'd like you to give the PR a thorough review -- to ensure it fixes what it is supposed to fix and also to learn:
- The main cause for the ["flakyness"](https://github.com/raiden-network/raiden/issues/389) of the test was an incorrect [Yoda-style](https://en.wikipedia.org/wiki/Yoda_conditions) boolean comparison in the [`blockalarm_for_settle`](https://github.com/raiden-network/raiden/pull/391/commits/3bac2c5158db69ef484f56bb3713f6e99e162973). I believe there is no reason for Yoda-boolean in python and we should avoid it, because it is harder to reason about it (the bug went unnoticed for several months).
- The flakyness of `test_settlement.py::test_settlement` was most likely caused by the sometimes late spawned `_settle` task (which could only then run successfully).
- Due to the skipped test in CI, we also missed a [breaking change for the arguments to `close`](https://github.com/raiden-network/raiden/commit/e82f7b3e5725f68d08c1791ff8a25515bfbd7232), which [was also fixed](https://github.com/raiden-network/raiden/pull/391/files#diff-9161c3328a41e5b779ebb4d7966b8abcR109).
- `test_settlement.py::test_settlement` [missed to properly check the postconditions](https://github.com/raiden-network/raiden/pull/391/files#diff-9161c3328a41e5b779ebb4d7966b8abcR142). Due to a [padding mismatch for the secret](https://github.com/raiden-network/raiden/pull/391/files#diff-9161c3328a41e5b779ebb4d7966b8abcR53) the [unlock call](https://github.com/raiden-network/raiden/pull/391/files#diff-9161c3328a41e5b779ebb4d7966b8abcR125) never succeeded before (i.e. there were assets lost). See also #395.
- I increased some of the pytest-timeouts to let the tests pass consistently here. I believe we should use timeouts **only** as a last resort to keep the test suite from locking up completely, but never to "speed things up".
- One of the other tests [affected by the fix]() already [suspected something fishy](https://github.com/raiden-network/raiden/pull/391/files#diff-e9ba3bf3caea82260af91c81e3782dbaL716) -- I suggest we should take the time to investigate odd behavior instead of coding around it :)
- ~~I added an [explicit context switch](https://github.com/konradkonrad/raiden/blob/99ba6e5066694af9f6bd0a025c778a1b86626106/raiden/tests/utils/blockchain.py#L44-L51) to the `wait_until_block` pattern, and applied it to all tests I touched. This [wasn't enough in all instances](https://github.com/raiden-network/raiden/pull/391/commits/fe7cfc2e49915f3dd5559a65551920c7db790d85) and I'd like to get your opinion on a best-practice for coroutine testing.~~

There are a couple of follow up issues:
- #393 -- issuing callbacks for lifecycle changes
- #394 -- channel proxy return values
- #395 -- hashlock/secret padding pitfall
- #396 -- debug tooling
- #397 -- review skip/xfail test markers

This fixes #390 :) :)

### CODE REVIEW

**Comment by LefterisJP** [comment]
> I added an explicit context switch to the wait_until_block pattern, and applied it to all tests I touched. This wasn't enough in all instances and I'd like to get your opinion on a best-practice for coroutine testing.

I remember we tried this in the past too. Both me and @hackaugusto had put this context switch in `wait_until_block()` but the improvement (or lack of) was also flaky and incosistent.

It would be a lot of work but can't we have a list of all background tasks that are awaiting for the context switch (would  depend on the test) and instead of waiting `X` number of seconds and hope they finish their task, continuously check for them doing what we expect them to do? 

And if they don't complete their tasks within a specified amount of time then the test would fail with a timeout.

File: raiden/channel.py [hunk 1]
```diff
@@ -410,6 +410,7 @@ def settled_block(self):
         return self._settled_block
 
     def set_opened(self, block_number):
+        # TODO: ensure the same callback logic as in set_settled
```
Comment (LefterisJP) [inline raiden/channel.py]: Add the issue number that tracks this here: https://github.com/raiden-network/raiden/issues/393
Question: Did you encounter a callback being invoked twice?

Comment (LefterisJP) [inline raiden/channel.py]: Add the issue number that tracks this here: https://github.com/raiden-network/raiden/issues/393
Question: Did you encounter a callback being invoked twice?

File: raiden/channel.py [hunk 2]
```diff
@@ -419,6 +420,7 @@ def set_opened(self, block_number):
             callback(block_number)
 
     def set_closed(self, block_number):
+        # TODO: ensure the same callback logic as in set_settled
```
Comment (LefterisJP) [inline raiden/channel.py]: same as above

File: raiden/channel.py [hunk 3]
```diff
@@ -428,16 +430,19 @@ def set_closed(self, block_number):
...
-        self._settled_block = block_number
-
-        for callback in self.callbacks_settled:
-            callback(block_number)
+            for callback in self.callbacks_settled:
+                callback(block_number)
 
     def query_settled(self):
-        return self.netting_channel.settled()
+        # FIXME: the if None: return 0 constraint should be ensured on the proxy side
```
Comment (LefterisJP) [inline raiden/channel.py]: Why this change? Isn't `settled()` initialized as 0 by default? Can the proxy call ever return `None`?

File: raiden/channel.py [hunk 4]
```diff
@@ -615,32 +620,33 @@ def channel_closed(self, block_number):
...
+
+            # wait for the settle event, it could be our transaction or our
+            # partner's
+            self.settle_event.wait(0.5)
+
+            if self.settle_event.is_set():
+                log.info('channel automatically settled')
+                return
+
+        if (self.external_state.closed_block != 0 and block_number >= (
```
Comment (LefterisJP) [inline raiden/channel.py]: very nice find konrad

File: raiden/tests/integration/test_settlement.py [hunk 5]
```diff
@@ -197,34 +214,37 @@ def test_settled_lock(tokens_addresses, raiden_network, settle_timeout, reveal_t
...
-def test_start_end_attack(token_address, raiden_chain, deposit):
+def test_start_end_attack(tokens_addresses, raiden_chain, deposit, reveal_timeout):
     """ An attacker can try to steal tokens from a hub or the last node in a
     path.
 
     The attacker needs to use two addresses (A1 and A2) and connect both to the
     hub H, once connected a mediated transfer is initialized from A1 to A2
     through H, once the node A2 receives the mediated transfer the attacker
-    uses the it's know secret and reveal to close and settles the channel H-A2,
+    uses the know secret and reveal to close and settles the channel H-A2,
```
Comment (LefterisJP) [inline raiden/tests/integration/test_settlement.py]: `know` -> `known`

File: raiden/tests/unit/test_channel.py [hunk 6]
```diff
@@ -710,18 +709,19 @@ def test_automatic_dispute(raiden_network, deposit, settle_timeout, reveal_timeo
     chain0 = app0.raiden.chain
     wait_until_block(chain0, chain0.block_number() + 1)
 
-    # wait until the settle timeout has passed
-    settle_expiration = chain0.block_number() + settle_timeout
+    assert channel0.external_state.closed_block != 0
+    assert channel1.external_state.closed_block != 0
+
+    # wait until the settle timeout has passed + time to mine/propagate the tx
+    settle_expiration = chain0.block_number() + settle_timeout + 3
```
Comment (LefterisJP) [inline raiden/tests/unit/test_channel.py]: Why the extra 3 blocks here?

File: raiden/tests/unit/test_channel.py [hunk 7]
```diff
@@ -710,18 +709,19 @@ def test_automatic_dispute(raiden_network, deposit, settle_timeout, reveal_timeo
...
-    gevent.sleep(1)
 
     # check that the channel is properly settled and that Bob's client
     # automatically called updateTransfer() to reflect the actual transactions
     assert channel0.external_state.settled_block != 0
     assert channel1.external_state.settled_block != 0
     assert token.balance_of(channel0.external_state.netting_channel.address.encode('hex')) == 0
-    assert token.balance_of(address0.encode('hex')) == initial_balance0 + deposit - 70 + 50
-    assert token.balance_of(address1.encode('hex')) == initial_balance1 + deposit + 70 - 50
+    total_alice = amount_alice1 + amount_alice2
```
Comment (LefterisJP) [inline raiden/tests/unit/test_channel.py]: Nice change, more readable test through the use of better variable names.

File: raiden/tests/utils/blockchain.py [hunk 8]
```diff
@@ -47,6 +47,8 @@ def wait_until_block(chain, block):
     curr_block = chain.block_number()
     while curr_block < block:
         curr_block = chain.next_block()
+    # ctx switch to allow background tasks to proceed:
+    gevent.sleep(1)
```
Comment (LefterisJP) [inline raiden/tests/utils/blockchain.py]: As you pointed out in the PR's comments we need to think about this.

File: raiden/channel.py (hunk 3 above)
Comment (konradkonrad) [inline raiden/channel.py]: yes, it did, at least on `tester` -- follow-up #394

File: raiden/channel.py (hunk 1 above)
Comment (konradkonrad) [inline raiden/channel.py]: No, but I did run into the `RuntimeError` in `set_settled` (which is there to prevent duplicate callbacks). It is a pretty defensive safe-guard and I believe the removal of https://github.com/raiden-network/raiden/pull/391/files/99ba6e5066694af9f6bd0a025c778a1b86626106#diff-08685f521533606c5888755314d415efL618 did fix `RuntimeError` I encountered.

File: raiden/channel.py (hunk 3 above)
Comment (konradkonrad) [inline raiden/channel.py]: see https://travis-ci.org/raiden-network/raiden/jobs/201871014#L477 of the demo PR #398

File: raiden/channel.py [hunk 9]
```diff
@@ -615,32 +620,33 @@ def channel_closed(self, block_number):
...
+            settled_block = self.external_state.query_settled()
+            if settled_block != 0:
+                self.external_state.set_settled(settled_block)
+                log.info('channel automatically settled')
+                return
+
+            try:
+                # TODO: to remove unecessary JSON-RPC calls, change to an
+                # async version of settle and stop polling if the
+                # settle_event was set
```
Comment (hackaugusto) [inline raiden/channel.py]: Humm, this TODO doesn't make much sense now, could you remove please?

File: raiden/channel.py [hunk 10]
```diff
@@ -615,32 +620,33 @@ def channel_closed(self, block_number):
...
+            except:
+                log.exception('Timedout while calling settle')
+
+            # wait for the settle event, it could be our transaction or our
+            # partner's
+            self.settle_event.wait(0.5)
+
+            if self.settle_event.is_set():
+                log.info('channel automatically settled')
+                return
```
Comment (hackaugusto) [inline raiden/channel.py]: since we are not looping anymore, we don't need to wait for the settle_event, the lines 639-645 can be removed (or we could wait for the logging info)

File: raiden/tests/utils/blockchain.py (hunk 8 above)
Comment (hackaugusto) [inline raiden/tests/utils/blockchain.py]: The `chain.block_number`  does context-switches, I don't think we need this sleep.

If we keep the line the sleep time should be a smaller, e.g. `gevent.sleep(0.000001)`.

File: raiden/tests/integration/test_settlement.py [hunk 11]
```diff
@@ -104,29 +106,44 @@ def test_settlement(raiden_network, settle_timeout, reveal_timeout):
...
+
+    channel1.external_state.netting_channel.unlock(
+        app1.raiden.address,
         [unlock_proof],
     )
 
-    settle_expiration = chain0.block_number() + settle_timeout
+    settle_expiration = chain0.block_number() + settle_timeout + 2
     wait_until_block(chain0, settle_expiration)
+    gevent.sleep(1)
```
Comment (hackaugusto) [inline raiden/tests/integration/test_settlement.py]: Why are we waiting for a second here? If it's for the event polling we can use the channel settle_event instead `channel0.settle_event.wait()`

Comment (konradkonrad) [inline raiden/tests/integration/test_settlement.py]: Good catch and thanks for the event waiting hint! I can't reproduce the issue I had anymore, so I'll purge the commit for this.

File: raiden/tests/unit/test_channel.py (hunk 6 above)
Comment (konradkonrad) [inline raiden/tests/unit/test_channel.py]: trial and error :( 
I changed the waiting model.

File: raiden/tests/unit/test_channel.py [hunk 12]
```diff
@@ -710,18 +709,23 @@ def test_automatic_dispute(raiden_network, deposit, settle_timeout, reveal_timeo
...
     settle_expiration = chain0.block_number() + settle_timeout
     wait_until_block(chain0, settle_expiration)
-    # manually call settle (automatic settling does not seem to work)
-    # TODO: ^ Find out why
-    channel1.external_state.settle()
-    gevent.sleep(1)
+
+    # the settle event must be set
+    assert channel0.settle_event.wait(timeout=60)
+    assert channel1.settle_event.wait(timeout=60)
```
Comment (konradkonrad) [inline raiden/tests/unit/test_channel.py]: @LefterisJP re: https://github.com/raiden-network/raiden/pull/391#discussion_r101255574

This passes now, but I had to use a timeout well in the range of 3 blocks to make it work. I expect the solution to be somewhat brittle :(

File: raiden/tests/integration/test_settlement.py [hunk 13]
```diff
@@ -104,29 +105,43 @@ def test_settlement(raiden_network, settle_timeout, reveal_timeout):
...
 
     # settle must be called by the apps triggered by the ChannelClose event,
     # and the channels must update it's state based on the ChannelSettled event
     assert channel0.external_state.settled_block != 0
     assert channel1.external_state.settled_block != 0
 
+    address0 = app0.raiden.address
+    address1 = app1.raiden.address
+
+    assert token.balance_of(address0.encode('hex')) == balance0 + deposit0 + amount
```
Comment (hackaugusto) [inline raiden/tests/integration/test_settlement.py]: `balance0 = token.balance_of(app0.raiden.address)` is queried without encoding the address while the assert is encoding, one of this calls seems wrong. I believe the token proxy will do the encoding of the address for you.

Comment (konradkonrad) [inline raiden/tests/integration/test_settlement.py]: good catch! I removed the superfluous hex encoding.

File: raiden/channel.py (hunk 9 above)
Comment (konradkonrad) [inline raiden/channel.py]: done

File: raiden/channel.py (hunk 10 above)
Comment (konradkonrad) [inline raiden/channel.py]: removed it

File: raiden/tests/utils/blockchain.py (hunk 8 above)
Comment (konradkonrad) [inline raiden/tests/utils/blockchain.py]: I removed this again after changing the waiting model to `wait_until_block` + `event.wait(timeout`

File: raiden/tests/integration/test_settlement.py (hunk 5 above)
Comment (konradkonrad) [inline raiden/tests/integration/test_settlement.py]: haha, close but no cigar. fixed it.

File: raiden/channel.py [hunk 14]
```diff
@@ -638,32 +643,24 @@ def channel_closed(self, block_number):
...
+
+            try:
+                self.external_state.settle()
+            except:
+                log.exception('Timedout while calling settle')
+
+        closed = self.external_state.closed_block != 0
+        after_settle_timeout = block_number >= (
+            self.external_state.closed_block + self.settle_timeout)
```
Comment (LefterisJP) [inline raiden/channel.py]: put the closing parentheses in its own line

File: raiden/channel.py (hunk 3 above)
Comment (LefterisJP) [inline raiden/channel.py]: Maybe before merging the PR let's add a link to issue 394 in the comment?

File: raiden/tests/integration/test_settlement.py [hunk 15]
```diff
@@ -197,34 +218,37 @@ def test_settled_lock(tokens_addresses, raiden_network, settle_timeout, reveal_t
...
     The intention is to make the hub transfer the token but for him to be
     unable to require the token A1.
     """
     amount = 30
 
-    token = token_address[0]
+    token = tokens_addresses[0]
     app0, app1, app2 = raiden_chain  # pylint: disable=unbalanced-tuple-unpacking
 
     # the attacker owns app0 and app2 and creates a transfer throught app1
```
Comment (LefterisJP) [inline raiden/tests/integration/test_settlement.py]: throught -> through

**Comment by LefterisJP** [comment]
HAve some minor comments that should be fixed but the PR is fine for merging with me after they are done.
